### PR TITLE
Loot parsing utilities

### PR DESCRIPTION
**Pull request by Cr3ativeZen**

### CODE REVIEW

File: apps/gameserver/include/Loot.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+
+struct LootGroup : public bango::processor::db_object<LootGroup>
+{
+    Group m_group;
+
+    unsigned int index() const { return m_group.m_loots_map.GetMaxKey() != 0 ? m_group.m_index : 0; }
+
+    void IsValueAllowed(LootInfo current) const;
```
Comment (lafreak) [inline apps/gameserver/include/Loot.h:64]: Signarue suggests the return value should be true/false, but is void

File: apps/gameserver/include/Loot.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    void AssignGroup(std::vector<uint32_t> values_from_bracket);
+
+    virtual void set(bango::processor::lisp::var param) override;
+};
+
+
+struct LootItemGroup : public bango::processor::db_object<LootItemGroup>
```
Comment (lafreak) [inline apps/gameserver/include/Loot.h:71]: Should the structures in this file be structs? Do we want to expose all fields and methods?

File: core/utils/test/main.cpp [hunk 3]
```diff
@@ -33,6 +34,43 @@ TEST(ClockTest, DISABLED_Difference)
...
+    interval_map<std::uint32_t> im;
+
+    //(group (index 120) (item (100 48 0) (200 1318 0) (350 48 0) (700 1318 0)(1000 1345 0)))
+    im.assign(0, 100, 48);
+    im.assign(100, 200, 1318);
+    im.assign(200, 350, 48);
+    im.assign(350, 700, 1318);
+    im.assign(700, 1000, 1345);
+
+    EXPECT_EQ(48, im[0]);
```
Comment (lafreak) [inline core/utils/test/main.cpp:48]: Gtest requires expected value to be second parameter, actual value first. This is important to get correct failure log.

File: core/utils/test/main.cpp [hunk 4]
```diff
@@ -33,6 +34,43 @@ TEST(ClockTest, DISABLED_Difference)
...
+TEST(IntervalMap, AssigningElements)
+{
+    interval_map<std::uint32_t> im;
+
+    //(group (index 120) (item (100 48 0) (200 1318 0) (350 48 0) (700 1318 0)(1000 1345 0)))
+    im.assign(0, 100, 48);
+    im.assign(100, 200, 1318);
+    im.assign(200, 350, 48);
+    im.assign(350, 700, 1318);
+    im.assign(700, 1000, 1345);
```
Comment (lafreak) [inline core/utils/test/main.cpp:46]: Needs testing wrong assignments such as:
im.assign(0, 5, 99);
im.assign(3, 8, 99);

Comment (lafreak) [reply core/utils/test/main.cpp:46]: What should we even expect in this scenario?

**Comment by lafreak** [comment]
Rename PR to better represent the changes - "Loot parsing utilities"

File: apps/gameserver/src/Loot.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        }
+        case A_ITEM:
+        {
+            while (!param.null()) 
+            {
+                auto values_from_bracket = GetValuesFromBrackets(param);
+                AssignGroup(values_from_bracket);
+            }
+            break;
+        }
```
Comment (lafreak) [inline apps/gameserver/src/Loot.cpp:27]:
```suggestion
        case IC_MONEY:
        case A_ITEM:
        {
            while (!param.null()) 
            {
                auto values_from_bracket = GetValuesFromBrackets(param);
                AssignGroup(values_from_bracket);
            }
            break;
        }
```

Both cases are indentical, merge to one?

File: apps/gameserver/include/Loot.h (hunk 1 above)
Comment (lafreak) [reply apps/gameserver/include/Loot.h:64]: I see, it checks whether the group is valid. How about rename to `Validate`?

File: apps/gameserver/src/Loot.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        loot_info = LootInfo(values_from_bracket.at(1), values_from_bracket.at(0), 1, values_from_bracket.at(2));
+    else
+        throw std::runtime_error("Invalid amount of values in brackets for group index: " + std::to_string(m_group.m_index));
+
+    IsValueAllowed(loot_info);
+    map.assign(max_key, values_from_bracket.at(0), loot_info);
+ }
+
+
+void LootItemGroup::IsValueAllowed(GroupInfo current) const
```
Comment (lafreak) [inline apps/gameserver/src/Loot.cpp:101]: We should catch those exceptions and report an error or handle the situation.

File: apps/gameserver/include/Loot.h (hunk 1 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/Loot.h:64]: done

File: core/processor/include/bango/processor/db.h [hunk 7]
```diff
@@ -76,7 +76,7 @@ namespace bango { namespace processor {
 
                     // BUG: Some configs have multiple types of rows for example InitNPC npc/gennpc. Filter?
                     if (m_db.find(temp.index()) == m_db.end()) 
-                        m_db.insert(std::make_pair(temp.index(), std::unique_ptr<T>(new T(temp) )));
+                        m_db.insert(std::make_pair(temp.index(), std::unique_ptr<T>(std::make_unique<T>(temp) )));
```
[the remark this answers is not in the document]
Comment (Cr3ativeZen) [reply core/processor/include/bango/processor/db.h:79]: done

File: core/utils/test/main.cpp (hunk 3 above)
Comment (Cr3ativeZen) [reply core/utils/test/main.cpp]: done

File: core/utils/test/main.cpp (hunk 4 above)
Comment (Cr3ativeZen) [reply core/utils/test/main.cpp:46]: in this case we should get 99 for interval of [0,8).
I added a lot of tests which will cover edge cases.

File: apps/gameserver/src/Loot.cpp (hunk 5 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Loot.cpp:27]: true, done.

File: apps/gameserver/include/Loot.h (hunk 2 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/Loot.h:71]: I think LootInfo, GroupInfo, Group, ItemGroup should stay as they are since they are just storing values and have some default constructors/ == operators.
I changed LootGroup and LootItemGroup to be classes and added getters.

File: apps/gameserver/include/Loot.h [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+struct ItemGroup
+{
+    ItemGroup() = default;
+    std::uint32_t m_index;
+    bango::utils::interval_map<GroupInfo> m_groups_map;
+};
+
+
+class LootGroup : public bango::processor::db_object<LootGroup>
```
Comment (lafreak) [inline apps/gameserver/include/Loot.h:58]: What's the point of changing from struct to class if we expose the group field anyway? I'd rather have all config structures declared the same way instead of mixing class/struct

File: apps/gameserver/src/Loot.cpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            while (!param.null()) 
+            {
+                auto values_from_bracket = GetValuesFromBrackets(param);
+                try
+                {
+                    AssignGroup(values_from_bracket);
+                }
+                catch(const std::exception& e)
+                {
+                    std::cerr << e.what() << '\n';
```
Comment (lafreak) [inline apps/gameserver/src/Loot.cpp:22]: Do we want to stop reading, report an error and close the app?

Comment (lafreak) [reply apps/gameserver/src/Loot.cpp:22]: I think that is the behavior of original kal server?

File: core/utils/include/bango/utils/interval_map.h [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+#include <iostream>
+
+namespace bango::utils
+{
+template<typename V>
+class interval_map
+{
+public:
+	V m_val_begin;
+	std::map<std::uint32_t, V> m_map;
```
Comment (lafreak) [inline core/utils/include/bango/utils/interval_map.h:14]: Do we need std::map or std::unordered_map is sufficient?

File: core/utils/include/bango/utils/interval_map.h [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    std::uint32_t max;
```
Comment (lafreak) [inline core/utils/include/bango/utils/interval_map.h:15]: Indent error

File: core/utils/include/bango/utils/interval_map.h [hunk 12]
```diff
@@ -0,0 +1,67 @@
+#include <assert.h>
+#include <map>
+#include <limits>
+#include <cstdint>
+#include <iostream>
+
+namespace bango::utils
+{
+template<typename V>
+class interval_map
```
Comment (lafreak) [inline core/utils/include/bango/utils/interval_map.h:10]: Why is it a class if we make everything public anyway? Would use struct instead.

On the other hand I also think that you do expose too much internals for the user. Is it for unit testing purposes? We should only unit test the contract that this utility provides, internal stuff should be hidden.

File: core/utils/include/bango/utils/interval_map.h [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+			auto [iterator, success] = m_map.insert_or_assign(keyEnd, currentKeyEndValue);
+            max = std::max(max, keyEnd);
+			endRemoveIterator = iterator;
+		}
+
+		if (beginRemoveIterator != m_map.end() && (endRemoveIterator == m_map.end() || beginRemoveIterator->first < endRemoveIterator->first))
+			m_map.erase(beginRemoveIterator, endRemoveIterator);
+	}
+
+    std::uint32_t GetMaxKey() const
```
Comment (lafreak) [inline core/utils/include/bango/utils/interval_map.h:51]: indent error

File: core/utils/test/main.cpp [hunk 14]
```diff
@@ -33,6 +34,354 @@ TEST(ClockTest, DISABLED_Difference)
...
+}
+
+TEST(interval_map, EmptyMapTest)
+{
+	interval_map<std::uint32_t> m;
+
+	EXPECT_EQ(m[10], 0);
+
+	EXPECT_EQ(m.m_map.size(), 0);
+	IsMapCanonic(m);
```
Comment (lafreak) [inline core/utils/test/main.cpp:61]: I don't think we need to check IsMapCanonic in all the tests, can't we make one test that checks the cannonicness?

File: core/utils/test/main.cpp [hunk 15]
```diff
@@ -33,6 +34,354 @@ TEST(ClockTest, DISABLED_Difference)
...
+}
+
+TEST(interval_map, TryToAddTheSameFirstMapElementAs_m_valBegin)
+{
+	interval_map<std::uint32_t> m;
+	m.assign(5, 10, 0);
+
+	EXPECT_EQ(m.m_map.size(), 0);
```
Comment (lafreak) [inline core/utils/test/main.cpp:69]: Instead of exposing `m_map`, you can create getter: `m.get_size()`;

Comment (lafreak) [reply core/utils/test/main.cpp:69]: I'm also not sure what size means in this context. Is it number of ranges? If so, why is it 0 if we clearly have one range [5,10]?

File: core/utils/include/bango/utils/interval_map.h (hunk 10 above)
Comment (Cr3ativeZen) [reply core/utils/include/bango/utils/interval_map.h:14]: interval_map requires std::map because upper_bound are only available for std::map.

File: core/utils/include/bango/utils/interval_map.h (hunk 11 above)
Comment (Cr3ativeZen) [reply core/utils/include/bango/utils/interval_map.h:15]: In my visual studio code it's correct. Seems like bug again

File: core/utils/test/main.cpp (hunk 15 above)
Comment (Cr3ativeZen) [reply core/utils/test/main.cpp:69]: as stated in testcase name: TryToAddTheSameFirstMapElementAs_m_valBegin

We are trying to add an element of range [5,10) which is the same value as valbegin.
When the interval_map is empty no matter what you trying to access by [] operator will return valbegin.
When you try to add any range with value == valbegin it's pointless when map is empty.

File: core/utils/include/bango/utils/interval_map.h (hunk 13 above)
Comment (Cr3ativeZen) [reply core/utils/include/bango/utils/interval_map.h:51]: same as above, buggy github

File: core/utils/test/main.cpp (hunk 14 above)
Comment (Cr3ativeZen) [reply core/utils/test/main.cpp:61]: We need to check if map is Canonic in every test case to make sure that assigns were done correctly.

File: core/utils/include/bango/utils/interval_map.h (hunk 12 above)
Comment (Cr3ativeZen) [reply core/utils/include/bango/utils/interval_map.h:10]: Everything is hidden under private now. I added a const getter for map access only.

**Review by lafreak — APPROVED** [review]

**Review by lafreak — APPROVED** [review]
